// @iwer/sem/lib/internal/components/locatable.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { SpatialEntityComponent, SpatialEntityComponentType, } from './component.js';
export class LocatableComponent extends SpatialEntityComponent {
    constructor(spatialEntity, initData) {
        super(spatialEntity);
        this.type = SpatialEntityComponentType.Locatable;
        this.position.copy(initData.position);
        this.orientation.copy(initData.orientation);
    }
    get position() {
        return this._spatialEntity.position;
    }
    get rotation() {
        return this._spatialEntity.rotation;
    }
    get orientation() {
        return this._spatialEntity.quaternion;
    }
    get initData() {
        return {
            position: {
                x: this.position.x,
                y: this.position.y,
                z: this.position.z,
            },
            orientation: {
                x: this.orientation.x,
                y: this.orientation.y,
                z: this.orientation.z,
                w: this.orientation.w,
            },
        };
    }
    toJSON() {
        return {
            locatable: this.initData,
        };
    }
}
