// @livekit/components-react/dist/hooks.mjs
import { R as e, a5 as t, a2 as i, S as u, D as n, U as o, w as r, b as c, u as p, i as P, a as d, E as l, c as T, a4 as g, F as m, d as S, n as k, m as v, G as A, _ as C, H as I, C as M, A as R, e as b, y as f, s as y, o as D, Z as E, Y as V, I as h, J as L, l as x, V as B, K as F, x as U, B as w, L as G, M as K, v as N, a1 as Q, a3 as W, a6 as q, N as H, O as J, f as O, g as X, p as Y, $ as Z, P as _, Q as $, k as j, h as z, W as ss, T as as, t as es, a0 as ts, r as is, X as us } from "./hooks-hQJmeINB.mjs";
import { u as os } from "./room-BP3SCCCd.mjs";
export {
  e as AgentEvent,
  t as MessagesEvent,
  i as SessionEvent,
  u as useAgent,
  n as useAudioPlayback,
  o as useAudioWaveform,
  r as useChat,
  c as useChatToggle,
  p as useClearPinButton,
  P as useConnectionQualityIndicator,
  d as useConnectionState,
  l as useDataChannel,
  T as useDisconnectButton,
  g as useEvents,
  m as useFacingMode,
  S as useFocusToggle,
  k as useGridLayout,
  v as useIsEncrypted,
  A as useIsMuted,
  C as useIsRecording,
  I as useIsSpeaking,
  os as useLiveKitRoom,
  M as useLocalParticipant,
  R as useLocalParticipantPermissions,
  b as useMediaDeviceSelect,
  f as useMediaDevices,
  y as useMultibandTrackVolume,
  D as usePagination,
  E as useParticipantAttribute,
  V as useParticipantAttributes,
  h as useParticipantInfo,
  L as useParticipantPermissions,
  x as useParticipantTile,
  B as useParticipantTracks,
  F as useParticipants,
  U as usePersistentUserChoices,
  w as usePinnedTracks,
  G as useRemoteParticipant,
  K as useRemoteParticipants,
  N as useRoomInfo,
  Q as useSequentialRoomConnectDisconnect,
  W as useSession,
  q as useSessionMessages,
  H as useSortedParticipants,
  J as useSpeakingParticipants,
  O as useStartAudio,
  X as useStartVideo,
  Y as useSwipe,
  Z as useTextStream,
  _ as useToken,
  $ as useTrackByName,
  j as useTrackMutedIndicator,
  z as useTrackToggle,
  ss as useTrackTranscription,
  as as useTrackVolume,
  es as useTracks,
  ts as useTranscriptions,
  is as useVisualStableUpdate,
  us as useVoiceAssistant
};
